### ISSUE TITLE
wal-push hangs after successfully uploading file to S3

### ISSUE THREAD

**Issue by bunsenmcdubbs** [wal-g/wal-g#471]
`wal-push` was stuck a WAL segment for ~4 days. I checked in S3 and the WAL segment was already there (from 2019-11-13). This seems to indicate that `wal-g` ran into a (deadlock?) issue after it successfully uploaded the WAL segment to S3 but before it was able to exit successfully.

| last_archived_wal | last_archived_time |  now | 
|-|- | -|
| 0000000200018B85000000F9  | 2019-11-13 22:17:33.417286+00 | 2019-11-17 19:39:47.86025+00 |

Running `strace` revealed that wal-g was waiting for a lock.

```
$ sudo ps --forest -o pid,tty,stat,time,cmd -g $(ps -o sid= -p 551)
  PID TT       STAT     TIME CMD
54501 ?        Ss   05:32:05 postgres: archiver process   archiving 0000000200018B85000000FA
  547 ?        S    00:00:00  \_ sh -c envdir /etc/wal-g.d/env /usr/local/bin/wal-g wal-push pg_wal/0000000200018B85000000FA
  551 ?        Sl   00:16:00      \_ /usr/local/bin/wal-g wal-push pg_wal/0000000200018B85000000FA
$ sudo strace -p 551
strace: Process 551 attached
futex(0x22c36c0, FUTEX_WAIT_PRIVATE, 0, NULL
```

Ultimately, I was able to send a `SIGHUP` signal to wal-g and then WAL archiving continued successfully. But I don't know why or how this happened... 

I've seen this kind of behavior on multiple different instances in the past. It has typically resolved on its own without intervention (maybe this was also going to resolve by itself as well).

**Settings and Configuration**
- Postgres version 10
- wal-g version 0.2.12
- Ubuntu 18.04

| Setting | Value |
| - | - |
| Parallelism | 18 | 
| Compression method | brotli |
| Remote destination | S3 |
| Using WALG_GPG_KEY_ID | true |

**Comment by x4m** [comment]
Hi @bunsenmcdubbs ! Thanks for the report!
I've fixed a similar-looking bug in 0.2.13 https://github.com/wal-g/wal-g/commit/3718a5ad6e371e9c16b9bf24113a437c67ee6bb3
But it could only reproduce if you have a very high rate of file archiving. Also, previously this bug was only reported for GCS (though there is nothing GCS-specific, GCS has more frequently occurring prerequisites).

If you will be able to observe WAL-G stuck for more than 30 minutes, can you please send it SIGABRT and place goroutine stacks here? Then I'll be able to confirm that it's a fixed bug or new one...

**Comment by bunsenmcdubbs** [comment]
Thanks for the prompt reply @x4m! That does look very suspicious. Our Postgres instances do experience very high write volume so that seems to fit the bill.

Beginner question: if I send `SIGABRT`, how do I find the diagnostic info that you are looking for? Will they show up in the Postgres logs or the terminal session where I run `kill`?

**Comment by bunsenmcdubbs** [comment]
I've encountered this issue again (on another Postgres instance). I got the following from the Postgres logs after sending `SIGABRT` to the wal-g process.
```
SIGABRT: abort
PC=0x968d21 m=0 sigcode=0

goroutine 0 [idle]:
runtime.futex(0x22c36c0, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7fff087eeb60, 0x919c72, ...)
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/sys_linux_amd64.s:531 +0x21
runtime.futexsleep(0x22c36c0, 0x7fff00000000, 0xffffffffffffffff)
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/os_linux.go:46 +0x4b
runtime.notesleep(0x22c36c0)
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/lock_futex.go:151 +0xa2
runtime.stopm()
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/proc.go:2016 +0xe3
runtime.findrunnable(0xc00006a000, 0x0)
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/proc.go:2487 +0x4dc
runtime.schedule()
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/proc.go:2613 +0x13a
runtime.park_m(0xc002cf6480)
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/proc.go:2676 +0xae
runtime.mcall(0x13aa5d0)
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/asm_amd64.s:299 +0x5b

goroutine 1 [sleep]:
time.Sleep(0x2faf080)
        /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/time.go:105 +0x14f
github.com/wal-g/wal-g/internal.(*BgUploader).Stop(0xc000466190)
        /home/travis/gopath/src/github.com/wal-g/wal-g/internal/bguploader.go:70 +0x2c
github.com/wal-g/wal-g/internal.HandleWALPush(0xc000466140, 0x7fff087efd36, 0x1f)
        /home/travis/gopath/src/github.com/wal-g/wal-g/internal/wal_push_handler.go:52 +0x2be
github.com/wal-g/wal-g/cmd/pg.glob..func7(0x227c240, 0xc0003a5c40, 0x1, 0x1)
        /home/travis/gopath/src/github.com/wal-g/wal-g/cmd/pg/wal_push.go:19 +0x69
github.com/wal-g/wal-g/vendor/github.com/spf13/cobra.(*Command).execute(0x227c240, 0xc0003a5c10, 0x1, 0x1, 0x227c240, 0xc0003a5c10)
        /home/travis/gopath/src/github.com/wal-g/wal-g/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/wal-g/wal-g/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x227d2e0, 0x1, 0x227c240, 0xc000397f78)
        /home/travis/gopath/src/github.com/wal-g/wal-g/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/wal-g/wal-g/vendor/github.com/spf13/cobra.(*Command).Execute(0x227d2e0, 0x914c70, 0xc000124058)
        /home/travis/gopath/src/github.com/wal-g/wal-g/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/wal-g/wal-g/cmd/pg.Execute()
        /home/travis/gopath/src/github.com/wal-g/wal-g/cmd/pg/pg.go:28 +0x2d
main.main()
        /home/travis/gopath/src/github.com/wal-g/wal-g/main/pg/main.go:8 +0x20

goroutine 6 [select]:
github.com/wal-g/wal-g/vendor/go.opencensus.io/stats/view.(*worker).start(0xc000163400)
        /home/travis/gopath/src/github.com/wal-g/wal-g/vendor/go.opencensus.io/stats/view/worker.go:154 +0x100
created by github.com/wal-g/wal-g/vendor/go.opencensus.io/stats/view.init.0
        /home/travis/gopath/src/github.com/wal-g/wal-g/vendor/go.opencensus.io/stats/view/worker.go:32 +0x57

rax    0xca
rbx    0x22c3580
rcx    0x968d23
rdx    0x0
rdi    0x22c36c0
rsi    0x80
rbp    0x7fff087eeb28
rsp    0x7fff087eeae0
r8     0x0
r9     0x0
r10    0x0
r11    0x286
r12    0x44
r13    0x4
r14    0xc0001d0f00
r15    0x0
rip    0x968d21
rflags 0x286
cs     0x33
fs     0x0
gs     0x0
2019-11-26 01:22:49 UTC [31171]: [15-1] LOG:  archive command failed with exit code 2
2019-11-26 01:22:49 UTC [31171]: [16-1] DETAIL:  The failed archive command was: envdir /etc/wal-g.d/env /usr/local/bin/wal-g wal-push pg_wal/00000002000115EC0000003A
```

**Comment by x4m** [comment]
Thanks @bunsenmcdubbs ! Yes, this is bug fixed in 3718a5a:
1. Postgres requested ```wal-push``` of WAL segment X
2. While performing upload, WAL-G noticed WAL segment X+1
3. Upload of X+1 failed due to some unknown reasons
4. But fail of X+1 did not decrement number of concurrent uploads
5. When upload of X finished, it found out there are still concurrent uploads. It's waiting for X+1, which will never be complete

**Comment by bunsenmcdubbs** [comment]
Awesome! I'll make sure I update shortly. Thanks for the help!
